src/schemata/date/DateFromIsoString.ts

```diff
@@ -0,0 +1,323 @@
+/**
+ * The Date parser (used in DateFromString) accepts different strings depending on
+ * runtime, and also accepts other formats like `February 29, 2022`.
+ *
+ * `DateFromIsoString` follows a subset of the [ECMAScript 2023 Language Date Time String
+ * Specification](https://tc39.es/ecma262/#sec-date-time-string-format).
+ *
+ * Notable features:
+ *
+ * - Requires `T` separator between date and time
+ * - Requires padded months, days, hours, minutes, and seconds
+ * - Can be configured to require a time, time and timezone offset (e.g. `Z` or `±05:00`) or
+ *   neither (default is require both).
+ * - Dates may contain years, months, and days; years and months; or years
+ * - Times may contain hours, minutes, seconds, and milliseconds; hours, minutes, and
+ *   seconds; or hours and minutes.
+ * - Expanded years are permitted (e.g. `+002022` instead of `2022`)
+ *
+ * @since 1.0.0
+ */
+
+import { pipe } from 'fp-ts/function'
+import * as PB from '../../PatternBuilder'
+import { make, SchemaExt } from '../../SchemaExt'
+import * as D from '../../schemables/WithDate'
+import { match } from '../../internal/match'
+
+/**
+ * E.g. `2022`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const year: PB.Pattern = pipe(PB.digit, PB.exactly(4))
+
+/**
+ * E.g. `+-002022`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const expandedYear: PB.Pattern = pipe(
+  PB.characterClass(false, '+', '-'),
+  PB.subgroup,
+  PB.then(pipe(PB.digit, PB.exactly(6))),
+)
+
+/**
+ * E.g. 01, 02, 03, 04, 05, 06, 07, 08, 09, 10, 11, 12
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const month: PB.Pattern = pipe(
+  PB.char('0'),
+  PB.then(pipe(PB.integerRange(1, 9), PB.subgroup)),
+  PB.subgroup,
+  PB.or(pipe(PB.integerRange(10, 12), PB.subgroup)),
+  PB.subgroup,
+)
+
+/**
+ * E.g. 01, 02, 03, 04, 05, ..., 30, 31
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const day: PB.Pattern = pipe(
+  PB.char('0'),
+  PB.then(pipe(PB.integerRange(1, 9), PB.subgroup)),
+  PB.subgroup,
+  PB.or(pipe(PB.integerRange(10, 31), PB.subgroup)),
+  PB.subgroup,
+)
+
+/**
+ * E.g. `YYYY-MM-DD`
+ *
+ * @since 1.0.0
+ * @category Pattern
+ */
+const yearMonthDay: PB.Pattern = pipe(
+  year,
+  PB.or(expandedYear),
+  PB.subgroup,
+  PB.then(PB.char('-')),
+  PB.then(month),
+  PB.then(PB.char('-')),
+  PB.then(day),
+)
```

Comment:
personal preference but these long chains of `.then` are why I added `sequence`

e.g. this could be, I think,

```ts
const yearMonthDay: PB.Pattern = PB.sequence(
  PB.subgroup(PB.oneOf(year, expandedYear)),
  PB.char('-'),
  month,
  PB.char('-'),
  day,
)
```

Reply:
(note this might require using a more specific type; I believe `sequence` only takes `Term`s)

Reply:
I'll try that out, thanks!